### ISSUE TITLE
[DataGridPro] if column type is singleSelect and column field is an array do a contains indexOf style filter

### ISSUE THREAD

**Issue by chrishawn** [mui-org/material-ui-x#3131]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

If my column data is an array i want the filter to do an indexOf into the array match. 

### Examples 🌈

If my column data is an array for example "tags"
and my column type is "singleSelect"
with valueOptions being ["tag1", "tag2", "tag3"]
i want to have when i pick one of those that the filter does an indexOf into the array and not an exact "is" match

### Motivation 🔦

I want the filtered matching data that contains that value while still allowing the filter to not be freeform to the user. This a feature that we use in [mui-datatables](https://github.com/gregnb/mui-datatables) but i would like to have that functionality when using DataGridPro

### Order ID 💳 (optional)

31818

**Comment by chrishawn** [comment]
or simply a "multiSelect" type field

**Comment by DanailH** [comment]
Hi, @chrishawn thank you for raising this. It looks like it might e a duplicate of https://github.com/mui-org/material-ui-x/issues/2514 🤔. You can set the `valueOptions` to be of different values than that what you would like to be displayed and then use the `valueFormatter` to map those values to whatever you want the user to see. It will also work for the filtering logic.

**Comment by chrishawn** [comment]
That is a different issue; in this case I can populate the options with the correct values. But the columns data would be like 

tag1
tag1,tag2
tag1,tag3
tag2,tag3

so if on if my singleSelect options of is tag1 it currently only gets the first row because it’s the only one that is an exact match

**Comment by alexfauquette** [comment]
The solution is probably to define a custom filter (see the [documentation](https://mui.com/components/data-grid/filtering/#create-a-custom-operator))

set the column `valueOptions = ["tag1", "tag2", "tag3"]` and define the filter logic `getApplyFilterFn` as follow

```js
getApplyFilterFn = (filterItem) => {
  if (filterItem.value == null || filterItem.value === "") {
    return null; // no value provided then no filtering
  }
  return ({ value }): boolean => {
    // The important line is the next one
    return value.includes(filterItem.value);
  };
}
```

This will allow users to select one tag from "tag1", "tag2", "tag3" to keep all rows containing this tag

By doing so you will override the default "is" filter which is not adapted to your case (default filters can be found here: https://github.com/mui-org/material-ui-x/blob/next/packages/grid/_modules_/grid/models/colDef/gridSingleSelectOperators.ts )

**Comment by chrishawn** [comment]
specifying an isAnyOf (preferably built in) filter operator on a column is reasonable, expecting someone to step in and augement the filter model and filter operator is a much bigger ask. i would assume most would be like me looking for something on the columns like

```
 {
    field: "tags",
    headerName: "Tags",
    type: "anyOfSelect",  // or "multiSelect"
    valueOptions: tags,
},
```

**Comment by m4theushw** [comment]
I'm closing since this request is the same of #2437 which is being addressed by #2874. You can subscribe to the linked issue to get notified when the new filter operator is available. For now, overriding the filter as explained in https://github.com/mui-org/material-ui-x/issues/3131#issuecomment-965535939 can be used as workaround.
